lib/Dialect/Comb/Transforms/LowerComb.cpp

```diff
@@ -0,0 +1,107 @@
+//===- LowerComb.cpp - Lower some ops in comb -------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/Comb/CombPasses.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace circt::comb;
+
+namespace circt {
+namespace comb {
+#define GEN_PASS_DEF_LOWERCOMB
+#include "circt/Dialect/Comb/Passes.h.inc"
+} // namespace comb
+} // namespace circt
+
+namespace {
+/// Lower truth tables to mux trees.
+struct TruthTableToMuxTree : public OpConversionPattern<TruthTableOp> {
+  TruthTableToMuxTree(MLIRContext *ctxt, bool assignToWire,
+                      uint64_t &symCounter)
+      : OpConversionPattern(ctxt), assignToWire(assignToWire),
+        symCounter(symCounter) {}
+
+private:
+  /// Get a mux tree for `inputs` corresponding to the given truth table. Do
+  /// this recursively by dividing the table in half for each input.
+  // NOLINTNEXTLINE(misc-no-recursion)
+  Value getMux(Location loc, OpBuilder &b, Value t, Value f,
+               ArrayRef<bool> table, Operation::operand_range inputs) const {
+    assert(!table.empty());
+    if (table.size() == 1)
+      return table.front() ? t : f;
+    assert(!inputs.empty());
+    assert(table.size() % 2 == 0);
+
+    size_t half = table.size() / 2;
+    Value if1 =
+        getMux(loc, b, t, f, table.drop_front(half), inputs.drop_front());
+    Value if0 =
+        getMux(loc, b, t, f, table.drop_back(half), inputs.drop_front());
+    return b.create<MuxOp>(loc, inputs.front(), if1, if0, true);
```

Comment:
Setting the `twoState` flag on the mux does not seem right to me. This way the truth table may end up being lowered to an if-else-tree, thus messing up the intended 'X behaviour.

Reply:
Does CIRCT currently have a conversion that would result in an if-else instead of a trinary op?

Problem is that I don't want to disable the mux canonicalizers. We don't care about preserving x-prop behavior in internal transformations. We just don't want SV xprop behavior in the output. Come to think of it, these are two separate issues...

Reply:
I couldn't replicate the 'if-else' output with your example, but I believe you. Yes, I agree not setting the flag is the safe thing to do for now.

Reply:
(your example gives me the following)

```sv
circt-opt --lower-comb --lower-seq-firrtl-to-sv=disable-reg-randomization --export-verilog test.mlir
// Generated by CIRCT unknown git version
module LutReg(  // test.mlir:2:1
  input  clk,   // test.mlir:2:19
         a,     // test.mlir:2:29
         b,     // test.mlir:2:37
  output x      // test.mlir:2:49
);

  reg reg0;     // test.mlir:4:13
  always @(posedge clk) // test.mlir:2:19
    reg0 <= a ? (b ? 1'h1 : 1'h0) : b ? 1'h0 : 1'h1;    // test.mlir:3:10, :4:13
  assign x = reg0;      // test.mlir:4:13, :5:5
endmodule
```